src/notify/notify.c

```diff
@@ -313,58 +336,271 @@ static int parse_argv(int argc, char *argv[]) {
-                help();
-                return -EINVAL;
+                if (n_arg_env > 0) {
+                        arg_env = strv_copy_n(argv + optind, n_arg_env);
+                        if (!arg_env)
+                                return log_oom();
+                }
+                if (!fdset_isempty(passed))
+                        log_warning("Warning: %u more file descriptors passed than referenced with --fd=.", fdset_size(passed));
         }
 
-        if (have_env && arg_action == ACTION_BOOTED)
-                log_warning("Notify message specified along with --booted, ignoring.");
+        case ACTION_BOOTED:
+                if (argc > optind)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "--booted takes no parameters, refusing.");
 
-        if (n_arg_env > 0) {
-                arg_env = strv_copy_n(argv + optind, n_arg_env);
-                if (!arg_env)
-                        return log_oom();
+                break;
+
+        case ACTION_FORK:
+                if (optind >= argc)
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "--fork requires a command to be specified, refusing.");
+
+                break;
         }
 
-        if (!fdset_isempty(passed))
-                log_warning("Warning: %u more file descriptors passed than referenced with --fd=.", fdset_size(passed));
+        if (have_env && arg_action != ACTION_NOTIFY)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "--ready, --reloading, --stopping, --pid=, --status=, --fd= may not be combined with --fork or --booted, refusing.");
 
         return 1;
 }
 
+static int on_notify_socket(sd_event_source *s, int fd, unsigned event, void *userdata) {
+        PidRef *child = ASSERT_PTR(userdata);
+        int r;
+
+        assert(s);
+        assert(fd >= 0);
+
+        _cleanup_free_ char *text = NULL;
+        _cleanup_(pidref_done) PidRef pidref = PIDREF_NULL;
+        r = notify_recv(fd, &text, /* ret_ucred= */ NULL, &pidref);
+        if (r == -EAGAIN)
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to receive notification message: %m");
+
+        if (!pidref_equal(child, &pidref)) {
+                log_warning("Received notification message from unexpected process " PID_FMT " (expected " PID_FMT "), ignoring.",
+                            pidref.pid, child->pid);
+                return 0;
+        }
+
+        const char *p = find_line_startswith(text, "READY=1");
+        if (!p || !IN_SET(*p, '\n', 0)) {
+                if (DEBUG_LOGGING) {
+                        _cleanup_free_ char *escaped = cescape(text);
+                        log_debug("Received notification message without READY=1, ignoring: %s", strna(escaped));
+                }
+        } else {
+                log_debug("Received READY=1, exiting.");
+                assert_se(sd_event_exit(sd_event_source_get_event(s), EXIT_SUCCESS) >= 0);
```

Comment:
I do not like assert_se() here. Rather than that, why not `return sd_event_exit(...);` ?
Or, like the child event handler in the below, how about to log the failure?

Reply:
(I proposed this, see rationales above)

Reply:
I know. But, I do not like your suggestion. Assertion should be used for arguments or ensuring logic of the current function, rather than called functions.

Reply:
Assertions are there to ensure some basic state of the program. In e.g. updatectl the same pattern for sd_event_exit is used. I don't quite follow here